grandpa: report equivocations

andresilva opened the pull request:
This PR adds equivocation reporting for GRANDPA. Equivocation detection is implemented in [finality-grandpa](https://github.com/paritytech/finality-grandpa) crate and exposed through `Environment` hooks (`prevote_equivocation` and `precommit_equivocation`).

The GRANDPA runtime module is extended with a `report_equivocation` method that takes an equivocation proof and a session membership proof. This method validates the membership proof using an abstract `KeyOwnerProofSystem` (historical session module currently) and the equivocation proof (different vote targets, valid signatures, etc.), if both are valid an offence is reported to an abstract `ReportOffence` type (offences module currently). Additionally, the method `submit_report_equivocation_extrinsic` creates and submits a signed transaction for reporting an equivocation.

From the node-side, when we want to report an equivocation we need to create the session membership proof. In particular, the membership proof can only be created at a block when the given validator is part of the current session. We figure out the last block at which the current set id is active in a given chain by looking at previously signalled authority changes and create the membership proof at that block. We then call `submit_report_equivocation_extrinsic` at the best block.

### Important

1. In order to report an equivocation we need to send a signed transaction, therefore we need to have some key available for this, taking into account that this is also the account that will be potentially rewarded. I created a new app-specific crypto ("fish") that should be used for all equivocation reporting tasks (this will also be used for BABE later on). Any comments on this?

1. The current implementation for generating membership proofs requires having state available for the block on which the session where the equivocation happened was valid. This is not problematic right now since we require validators to run as archive. A future implementation could leverage offchain workers for generating full membership proof tries when a session ends and storing them locally.

**TODO**:
- [x] Publish finality-grandpa 0.9.1 (includes fixes for building under no-std)
- [X] Fix execution context of `submit_report_equivocation_extrinsic` call (needs to be offchain context but it's not easy to use that now outside of actual offchain workers)
- [X] Abstract srml/grandpa reporting logic to trait (so that we can implement it for `()` to disable reporting, needed to integrate in node-template)
- [x] Merge #4200 and drop cherry-picked changes
- [x] Cleanup
- [x] Tests
- [x] Docs

polkadot companion: paritytech/polkadot#1000

burdges commented:
We only need the key signing a report transaction to be an account key so that it can pay some fee, right?  We've no existing mechanism for actions signed by session keys to change their controller account fees yet?  We could maybe permit free reports by the grandpa ed25519 key if the deduplication works well enough, not so universally applicable though.  We can chat about this when you like.  :)

kianenigma commented on bin/node/executor/tests/submit_transaction.rs:
```diff
@@ -15,7 +15,7 @@
 
 use node_runtime::{
-	Call, Executive, Indices, Runtime, SubmitTransaction, UncheckedExtrinsic,
+	Call, Executive, Indices, Runtime, SubmitImOnlineTransaction, UncheckedExtrinsic,
```
Whoever loses the race of merging first needs to update this as they will conflict :D 
https://github.com/paritytech/substrate/pull/4517/files#diff-81c5ce5884e59ad2b5de584efa307f0aR34

kianenigma replied:
You lost :D

andresilva replied:
We both did actually :D

vkandy commented:
Would you consider allowing custom implementation of `GrandpaEquivocationOffence` to be passed to `report_offence`? We want to be able to change the formula for calculating the slash fraction. If we can avoid cloning the pallet just to be able to change the formula that would be perfect.

andresilva commented:
@vkandy Yeah I think that's doable, I'll add support for this.

vkandy commented:
@andresilva thank you!

andresilva commented:
I will add some more tests but I think this is ready for a review. Please read the PR description wrt the reporting requiring an app-specific account with funds and also @burdges comments on it. The easiest way to implement this is using appcrypto and my idea is that this key will be shared for any kind of reporting/fishermen activity. I'd appreciate opinions on this.

andresilva commented on primitives/core/src/lib.rs:
```diff
@@ -108,8 +108,11 @@ impl ExecutionContext {
+			// Enable keystore and transaction pool by default for offchain calls.
+			OffchainCall(None) => [
+				offchain::Capability::Keystore,
+				offchain::Capability::TransactionPool,
```
@tomusdrw Does this make sense? It would be hard for me to specify my own context in GRANDPA code since it requires providing `Externalities` (I think it would defeat the purpose of using offchain for this).

andresilva commented on frame/grandpa/src/lib.rs:
```diff
@@ -189,10 +211,81 @@ decl_module! {
+		fn report_equivocation(
+			origin,
+			equivocation_proof: EquivocationProof<T::Hash, T::BlockNumber>,
+			// key_owner_proof: <T::HandleEquivocation as equivocation::HandleEquivocation<T>>::KeyOwnerProof,
```
I don't know how to fix this. @bkchr mentioned that this is probably some rustc bug.

andresilva replied:
This is what I did in the initial implementation, the problem is that it requires always having a `KeyOwnerProofSystem` in order to instantiate GRANDPA (e.g. node-template doesn't use any of this), so I refactored all that into the `HandleEquivocation` trait that we can implement for `()` in order to equivocation reporting. We can't implement `KeyOwnerProofSystem` for `()` we'd need something like struct `NoopKeyOwnerProofSystem<Key, Proof, Id>(PhantomData<(Key, Proof, Id)>)`.

andresilva replied:
No longer an issue since I moved `KeyOwnerProof` to the module trait and there's no extra indirection now.

dud1337 commented:
Any updates on the bug fix?

andresilva commented:
@dud1337 I believe the overall approach is correct and good to go. There's some minor grumbles with the PR (I commented in the PR about them) and a question regarding the approach for defining the reporting account (using an app specific key). This is missing reviews and some extra tests.

kianenigma commented on bin/node/primitives/src/lib.rs:
```diff
@@ -62,3 +62,32 @@ pub type Header = generic::Header<BlockNumber, BlakeTwo256>;
+
+	/// Key type for the reporting module. Used for reporting BABE and GRANDPA
+	/// equivocations.
+	pub const KEY_TYPE: KeyTypeId = KeyTypeId(*b"fish");
```
comes from fisherman?

andresilva replied:
Yeah, initially I set it to "rprt" but thought this was cooler :P

kianenigma replied:
"fish" is cooler 🕶

kianenigma commented on client/finality-grandpa/src/environment.rs:
```diff
@@ -924,22 +929,133 @@ where
+/// generates a session membership proof of the offender and then submits an
+/// extrinsic to report the equivocation. In particular, the session membership
+/// proof must be generated at the block at which the given set was active which
+/// isn't necessarily the best block if there are pending authority set changes.
```
when and how do we have pending authorities? is this a reflection of the queued validators in session module?

rphmeier replied:
@kianenigma From the runtime perspective, we don't have pending authorities. However, GRANDPA set changes are pending until finalized. When finality lags we can have a DAG of unfinalized changes. In normal circumstances a GRANDPA set-change is not pending for more than a few seconds.

kianenigma commented on frame/grandpa/src/equivocation.rs:
```diff
@@ -0,0 +1,270 @@
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//!
```
```suggestion
```

kianenigma commented on frame/grandpa/src/lib.rs:
```diff
@@ -183,10 +205,81 @@ decl_module! {
+		///
+		/// FIXME: I have no clue about the weight (but we're checking two
+		/// ed25519 signatures).
+		#[weight = SimpleDispatchInfo::FixedOperational(10_000_000)]
```
Can be ignored for now but this value is 1/10 of the block limit. I don't think this is your intention?

andresilva replied:
I take that's way too much :relaxed: We should definitely come up with some sensitive value before merging (not necessarily too accurate for now).

andresilva replied:
Moved validation to a `SignedExtension` so the weight is now 0.

mxinden commented on client/finality-grandpa/src/environment.rs:
```diff
@@ -924,22 +929,133 @@ where
 	) {
 		warn!(target: "afg", "Detected prevote equivocation in the finality worker: {:?}", equivocation);
-		// nothing yet; this could craft misbehavior reports of some kind.
+		match report_equivocation(
```
Why not have `report_equivocation` a method within an `Environment` `impl`? Would save us on passing the first 3 arguments.

andresilva replied:
Done.

mxinden commented on client/finality-grandpa/src/environment.rs:
```diff
@@ -924,22 +929,133 @@ where
 		warn!(target: "afg", "Detected prevote equivocation in the finality worker: {:?}", equivocation);
-		// nothing yet; this could craft misbehavior reports of some kind.
+		match report_equivocation(
+			&self.client.clone(),
```
```suggestion
			&self.client,
```

mxinden commented on client/finality-grandpa/src/environment.rs:
```diff
@@ -924,22 +929,133 @@ where
 		warn!(target: "afg", "Detected precommit equivocation in the finality worker: {:?}", equivocation);
-		// nothing yet
+		match report_equivocation(
+			&self.client.clone(),
```
```suggestion
			&self.client,
```

mxinden commented on client/finality-grandpa/src/environment.rs:
```diff
@@ -580,6 +584,7 @@ where
 	Block: 'static,
 	B: Backend<Block>,
 	C: crate::ClientForGrandpa<Block, B> + 'static,
+	C::Api: GrandpaApi<Block>+ SessionMembership<Block, Error = sp_blockchain::Error>,
```
```suggestion
	C::Api: GrandpaApi<Block> + SessionMembership<Block, Error = sp_blockchain::Error>,
```

mxinden commented on client/finality-grandpa/src/lib.rs:
```diff
@@ -615,6 +617,7 @@ pub fn run_grandpa_voter<Block: BlockT, BE: 'static, C, N, SC, VR>(
 	NumberFor<Block>: BlockNumberOps,
 	DigestFor<Block>: Encode,
 	C: ClientForGrandpa<Block, BE> + 'static,
+	C::Api: GrandpaApi<Block>+ SessionMembership<Block, Error = sp_blockchain::Error>,
```
```suggestion
	C::Api: GrandpaApi<Block> + SessionMembership<Block, Error = sp_blockchain::Error>,
```

mxinden commented on client/finality-grandpa/src/lib.rs:
```diff
@@ -710,6 +713,7 @@ where
 	Block: BlockT,
 	B: Backend<Block> + 'static,
 	C: ClientForGrandpa<Block, B> + 'static,
+	C::Api: GrandpaApi<Block>+ SessionMembership<Block, Error = sp_blockchain::Error>,
```
```suggestion
	C::Api: GrandpaApi<Block> + SessionMembership<Block, Error = sp_blockchain::Error>,
```

mxinden commented on client/finality-grandpa/src/lib.rs:
```diff
@@ -896,6 +899,7 @@ where
 	NumberFor<Block>: BlockNumberOps,
 	SC: SelectChain<Block> + 'static,
 	C: ClientForGrandpa<Block, B> + 'static,
+	C::Api: GrandpaApi<Block>+ SessionMembership<Block, Error = sp_blockchain::Error>,
```
```suggestion
	C::Api: GrandpaApi<Block> + SessionMembership<Block, Error = sp_blockchain::Error>,
```

rphmeier commented on bin/node/runtime/src/lib.rs:
```diff
@@ -845,6 +869,16 @@ impl_runtime_apis! {
 
+	impl sp_session::SessionMembership<Block> for Runtime {
+		fn generate_session_membership_proof(
+			session_key: (sp_core::crypto::KeyTypeId, Vec<u8>),
```
Shouldn't the `Vec<u8>` here be a concrete key type?

andresilva replied:
The session module only deals with opaque key types.

rphmeier replied:
That is true, but the implementation of `KeyOwnerProofSystem` is written as

```rust
impl<T: Trait, D: AsRef<[u8]>> frame_support::traits::KeyOwnerProofSystem<(KeyTypeId, D)>
	for Module<T>
```

I did it this way so that modules can avoid using `Vec`ly-typed keys with the `KeyOwnerProofSystem`, because the underlying bytes are the same. It's more understandable and more efficient to avoid `Vec`.

rphmeier replied:
https://github.com/paritytech/polkadot/pull/951

andresilva replied:
Updated to do the same thing.

rphmeier commented on client/finality-grandpa/src/environment.rs:
```diff
@@ -924,22 +929,133 @@ where
+	let membership_proof = match client
+		.runtime_api()
+		.generate_session_membership_proof(
+			&BlockId::Hash(current_set_latest_hash),
```
Hopefully with the work of @drahnr on #3722  we would be able to call all of this stuff at the most recent block instead, as the offchain DB would hold the set membership tries.

rphmeier commented on primitives/finality-grandpa/src/lib.rs:
```diff
@@ -157,6 +164,241 @@ impl<N: Codec> ConsensusLog<N> {
+
+/// Verifies the equivocation proof by making sure that both votes target
+/// different blocks and that its signatures are valid.
+pub fn check_equivocation_proof<H, N>(report: &EquivocationProof<H, N>) -> Result<(), ()>
```
I prefer it the way it is, because `self.check()` just feels a bit weird to me organizationally.

rphmeier commented on primitives/session/src/lib.rs:
```diff
@@ -44,6 +47,33 @@ sp_api::decl_runtime_apis! {
 	}
+
+	/// Historical session membership runtime api.
+	pub trait SessionMembership {
```
Maybe we can just hide this in a GRANDPA runtime API to avoid leaking runtime-internal details about the proof. After all, I don't think node-side code will inspect this proof deeply.

andresilva replied:
Done.

rphmeier commented on frame/session/src/historical.rs:
```diff
@@ -253,26 +254,16 @@ impl<T: Trait> ProvingTrie<T> {
 
+	// The session index of a given membership proof and the validator count for
+	// that session.
+	type ExtraData = (SessionIndex, ValidatorCount);
```
This approach seems difficult to reconcile with the trait system. The property that users of `KeyOwnerProofSystem` need would be something like `Proof::ExtraData: Has<SessionIndex> + Has<ValidatorCount>` which can only be expressed via `frunk`'s `HList` ATM. Either that or you require `ExtraData = (SessionIndex, ValidatorCount)` which feels fragile.

The way it was done for parachain-equivocations in Polkadot was to add a trait and then you can just ensure `Proof: GetSessionIndex + GetValidatorCount`

andresilva replied:
Implemented this now and reverted the changes to the historical module.

andresilva commented on frame/grandpa/src/lib.rs:
```diff
@@ -359,6 +452,22 @@ impl<T: Trait> Module<T> {
+	/// Only useful in an offchain context.
+	pub fn submit_report_equivocation_extrinsic(
+		equivocation_proof: EquivocationProof<T::Hash, T::BlockNumber>,
+		key_owner_proof: <T::HandleEquivocation as equivocation::HandleEquivocation<T>>::KeyOwnerProof,
```
We would have to parameterize `EquivocationProof` on `K` (`K = <T::HandleEquivocation as equivocation::HandleEquivocation<T>>::KeyOwnerProof`), which would make it impossible to define the runtime API since we don't have access to that type there (https://github.com/paritytech/substrate/pull/3868/files#diff-233fff7c89a6b4ca96ca4ee31c62088cR483).

andresilva commented on frame/grandpa/src/lib.rs:
```diff
@@ -183,10 +205,81 @@ decl_module! {
+
+			// validate the membership proof and extract session index and
+			// validator set count of the session that we're proving membership of
+			let (offender, session_index, validator_set_count) =
```
This is done by checking against the `set_id -> session_id` mapping in the module's storage.

andresilva commented:
> the weight of the call to report_equivocation may be set to 0 with taking further steps to avoid DoS.

Moving this to a `SignedExtension` just feels like moving the problem somewhere else, i.e. instead of DoSing the runtime you DoS the transaction pool. IMO there is no getting around finding a proper weight for the method and users paying the fees in order to report.

montekki commented:
> Moving this to a `SignedExtension` just feels like moving the problem somewhere else, i.e. instead of DoSing the runtime you DoS the transaction pool. IMO there is no getting around finding a proper weight for the method and users paying the fees in order to report.

Could you describe a possible attack vector? The way I understand the inner workings of the tx pool is following:

1) all TXs enter the pool with a public api `submit_at` that among other things calls `verify` on the transactions.
2) In any case for each single TX we will end up in `verify_one` where `validate_transaction` from the `frame/executive` will be called.
3) `validate_transaction` among other things will call `SignedExtension` validation logic on this TX.
4) If that fails the transaction is not let into the pool.

Now, I am wondering how the DoS vector would look like here. For instance Eve could send a bunch of transactions that do not pass the `SignedExtension` validation phase. By "a bunch" I mean enough to make their discarding a) computaionally heavy b) bandwidth-heavy. In that case we have a problem with all TXs validated in the same fashion (not only this one case) unless we take some further steps, like banning Eve for a bad reputation or something.

A side note: the discussion on this validation pipeline regularly pops up so we probably should document it somewhere in detail, discussing all invariants.

rphmeier commented:
I guess the DoS is that you can have an unlimited number of accounts submit the same (valid) report. But it should be easier to implement restrictions on the number of transactions in the pool so that most, if not all, duplicates are rejected. Or, in practice a block author will include this TX as a local transaction.

montekki commented:
> I don't think we want duplicate or invalid reports successfully going into, but being rejected by the runtime.

If this is ok: say two honest fishermen detected the same double vote, you would then need to somehow pick one and only that one gets a reward. That might not be very motivating for the fishermen?

andresilva commented on primitives/finality-grandpa/src/lib.rs:
```diff
@@ -157,6 +164,241 @@ impl<N: Codec> ConsensusLog<N> {
+
+			return Ok(());
+		};
+	}
```
I am unsure of what to do here, the reason I'm using a macro is because the two types we're checking even though have the same fields are completely separate types. I could add some methods to `EquivocationProof` to extract most of the data and then check it without the macro but I think this would add too much boilerplate (e.g. `fn first_hash(), first_number(), first_message(), ...`) and on each of these implementations we'd have the same code just matching for a `Prevote` or a `Precommit`. Since checking the proof is the only thing we're doing with it I think it is not worth to add that boilerplate?

I have removed the clones though :)

andresilva commented on primitives/finality-grandpa/src/lib.rs:
```diff
@@ -157,6 +164,241 @@ impl<N: Codec> ConsensusLog<N> {
+	/// Returns the authority id of the equivocator.
+	pub fn offender(&self) -> &AuthorityId {
+		match self {
+			Equivocation::Prevote(ref equivocation) => &equivocation.identity,
```
That doesn't work because the inner equivocations have different types (same problem as above with the macro thing).

andresilva commented on client/finality-grandpa/src/environment.rs:
```diff
@@ -924,22 +929,133 @@ where
 	) {
 		warn!(target: "afg", "Detected prevote equivocation in the finality worker: {:?}", equivocation);
-		// nothing yet; this could craft misbehavior reports of some kind.
+		match report_equivocation(
```
Done.

andresilva commented:
I have addressed all review comments and also addressed some things discussed on chat:
- Moved `KeyOwnerProofSystem` out of `HandleEquivocation` and into the main module trait;
- Moved equivocation report validation to a `SignedExtension`. The weight of the call is now 0 so the reporting account doesn't need to be funded;
- Moved key ownership proof runtime API method to the grandpa runtime API and the proof itself is exposed as an `OpaqueKeyMembershipProof` (at the runtime API boundary we don't know the exact type of the proof so we need to deal with an encoded version).

rphmeier commented on bin/node/runtime/src/lib.rs:
```diff
@@ -569,6 +574,23 @@ impl pallet_authority_discovery::Trait for Runtime {}
+
+	type HandleEquivocation = pallet_grandpa::EquivocationHandler<
+		Self::KeyOwnerIdentification,
+		TransactionSubmitterOf<node_primitives::report::ReporterId>,
```
modularity here is really good 👍

rphmeier commented on bin/node/runtime/src/lib.rs:
```diff
@@ -786,6 +810,31 @@ impl_runtime_apis! {
+		}
+
+		fn generate_key_ownership_proof(
+			authority_id: fg_primitives::AuthorityId,
```
We probably want to provide `set_id` as a parameter here, since @drahnr will be changing `Historical::prove` to accept `SessionIndex`. I believe we have a `SetId -> SessionIndex` lookup somewhere.

andresilva replied:
Yeah we have a map `SetId -> Session` in the grandpa pallet which maps to the latest session in the set. I'll update this API (and keep the parameter unused for now).

andresilva commented on bin/node/primitives/src/lib.rs:
```diff
@@ -62,3 +62,34 @@ pub type Header = generic::Header<BlockNumber, BlakeTwo256>;
+
+	/// Key type for the reporting module. Used for reporting BABE and GRANDPA
+	/// equivocations.
+	pub const KEY_TYPE: KeyTypeId = KeyTypeId(*b"fish");
```
Agreed will add to follow-up PR.

andresilva commented on primitives/session/src/lib.rs:
```diff
@@ -46,6 +49,32 @@ sp_api::decl_runtime_apis! {
+
+impl MembershipProof {
+	/// Returns a session this proof was generated for.
+	pub fn session(&self) -> SessionIndex {
```
Yeah they don't.

andresilva commented on primitives/finality-grandpa/src/lib.rs:
```diff
@@ -157,6 +164,242 @@ impl<N: Codec> ConsensusLog<N> {
+{
+	localized_payload_with_buffer(round, set_id, message, buf);
+
+	#[cfg(not(feature = "std"))]
```
native/wasm version, tbh I already wrote this specific function some time ago I no longer remember why this was needed. Will check again and change in follow-up if unnecessary.

andresilva commented on frame/grandpa/src/lib.rs:
```diff
@@ -351,14 +429,35 @@ impl<T: Trait> Module<T> {
+	/// extrinsic with a call to `report_equivocation` with any reporting keys
+	/// available in the keystore and will push the transaction to the pool.
+	/// Only useful in an offchain context.
+	pub fn submit_report_equivocation_extrinsic(
```
It means that the transaction wasn't sent, either there are no signing accounts available locally for this or the given chain doesn't have equivocation reporting enabled at all. I can add this to the docs.

andresilva commented on frame/grandpa/src/lib.rs:
```diff
@@ -351,14 +429,35 @@ impl<T: Trait> Module<T> {
+		T::HandleEquivocation::submit_equivocation_report(equivocation_proof, key_owner_proof)
+			.ok()?;
+
+		Some(())
```
👍

andresilva commented on frame/grandpa/src/lib.rs:
```diff
@@ -351,14 +429,35 @@ impl<T: Trait> Module<T> {
 		}
+
+		// NOTE: initialize first session of first set. this is necessary
+		// because we only update this `on_new_session` which isn't called
```
Yeah initially the code here was only on genesis builder, I moved it here and didn't update the comment.

andresilva commented on frame/grandpa/src/lib.rs:
```diff
@@ -183,11 +223,49 @@ decl_module! {
+		/// against the extracted offender. If both are valid, the offence
+		/// will be reported.
+		///
+		/// Since the weight is 0 in order to avoid DoS pre-validation is implemented in a
```
This is missing some commas, should read as:
"Since the weight is 0, in order to avoid DoS, pre-validation is implemented in a..."

andresilva commented on frame/grandpa/src/equivocation.rs:
```diff
@@ -0,0 +1,406 @@
+#[repr(u8)]
+pub enum ReportEquivocationValidityError {
+	/// The proof provided in the report is not valid.
+	InvalidEquivocationProof = 1,
```
Where should they go then? Or do I just look for the next free `u8`?

andresilva commented on frame/grandpa/src/equivocation.rs:
```diff
@@ -0,0 +1,406 @@
+		_info: &DispatchInfoOf<Self::Call>,
+		_len: usize,
+	) -> TransactionValidity {
+		let (equivocation_proof, key_owner_proof) = match call.is_sub_type() {
```
Initially this was checked in the dispatchable and there was no signed extension. We want to avoid users being able to spam the chain with fake equivocation reports, one way to do it is to have a proper weight here and make sure that users pay for the transaction (this was what I did initially). But this requires that the reporting account has enough balance to pay these fees. So instead we decided to make the transaction free and validation happens in the signed extension, if the signed extension is not added then this doesn't work properly.

AurevoirXavier commented on bin/node/cli/Cargo.toml:
```diff
@@ -85,8 +85,9 @@ pallet-balances = { version = "2.0.0-dev", path = "../../../frame/balances" }
-pallet-staking = { version = "2.0.0-dev",  path = "../../../frame/staking" }
+pallet-authority-discovery = { version = "2.0.0-dev", path = "../../../frame/authority-discovery" }
+pallet-staking = { version = "2.0.0-dev", path = "../../../frame/staking" }
+pallet-grandpa = { version = "2.0.0-dev", path = "../../../frame/grandpa" }
```
This should be in `dev-dependencies`
